QF-465 Exit interviews journeys

techtoniq opened the pull request.

Commit:
Initial Start of Exit Survey

techtoniq committed:
Happy and unhappy user journeys without database persistence

CodeWolf86 commented on src/SFA.DAS.ApprenticeFeedback.Web/Startup/AuthenticationStartup.cs:
```diff
@@ -53,11 +53,13 @@ public static class AuthenticationStartup
 
             services.AddRazorPages(options =>
             {
+                /*
```
Know it's draft but making sure this is removed 😄

CodeWolf86 commented on src/SFA.DAS.ApprenticeFeedback.Web/Pages/ExitInterview/Question4.cshtml.cs:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        public IActionResult OnGet([FromServices] AuthenticatedUser user)
+        {
```
Potential bug here, whereby if on the check your answers page they select back, then on question 4 they select back. the session will still have the CheckAnswers value set to true so on forward would skip question 4.

Unsure how best to go about that, other than not having a back link on the check your answers page?

techtoniq replied:
The user could use the back button on the browser to achieve the same, so I'm not sure removing the back link is the answer?

CodeWolf86 commented on src/SFA.DAS.ApprenticeFeedback.Web/Pages/ExitInterview/Question1.cshtml:
```diff
@@ -0,0 +1,47 @@
+                    </legend>
+
+                    <div id="options-hint" class="govuk-hint">
+                        You did not complete the apprenticeship and or your end point assessment.
```
I think this part is the wrong style compared to the prototype

CodeWolf86 reviewed (changes requested):
All looks good, few minor comments so far. Main one is the validation I don't think looks right against the radio question. Take a look here https://design-system.service.gov.uk/components/error-message/

techtoniq committed:
Tweaks to support API integration

techtoniq committed:
Rename exit interview to exit survey and start integrating with outer api

techtoniq committed:
Undo dev comment out of auth

techtoniq committed:
Debug launch settings with apprentice target id

techtoniq committed:
Fix bug in RestEase API definition

techtoniq committed:
Change start path to accept apprentice feedback target id

techtoniq committed:
Refactor radios to checkboxes on question 3
Add client side validation for question 3
Update content and answers to match latest prototype

techtoniq committed:
Fix code smells

techtoniq committed:
Remove tabindex from validation error panel to remove yellow focus border on page load

techtoniq committed:
Make the checkboxes behave like a radio group

techtoniq committed:
Rename "other" to "none" and fix validation

techtoniq committed:
Tweak the shared page header to be specific for Exit Survey

techtoniq committed:
Fixing SonarCloud fussiness

techtoniq committed:
Fix validation tag helper class merging bug

techtoniq committed:
Fix back navigation to use feedback target ID

techtoniq committed:
Validation UI tweaks

CodeWolf86 commented on src/SFA.DAS.Apprentice.Feedback.Infrastructure/Session/ExitInterviewSessionService.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        private readonly ISessionService _sessionService;
+
+        private const string _sessionKey = "Exit_Interview_Context";
```
Minor point, tweak to survey?

CodeWolf86 commented on src/SFA.DAS.ApprenticeFeedback.Domain/Models/ExitInterview/ExitInterviewContext.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace SFA.DAS.ApprenticeFeedback.Domain.Models.ExitSurvey
+{
+    public class ExitSurveyContext
```
Just need to tweak the file name to keep the class and file consistent

CodeWolf86 commented on src/SFA.DAS.ApprenticeFeedback.Web/Pages/ExitSurvey/IncorrectReason.cshtml.cs:
```diff
@@ -0,0 +1,85 @@
+            SaveContext();
+
+            // Save 
+            var request = new PostSubmitExitSurvey
```
Possibly a design question. 

Should Incorrect Reason upon submission also go to check your answers? I don't think it does on the prototype because it's so simple. "no i did complete or i haven't withdrawn"  "ok why... answer... submit"

just a thought I had but maybe one to ignore!

techtoniq committed:
Address feedback

techtoniq committed:
QF-666 unhappy path for "already completed apprenticeship"

techtoniq committed:
Prevent some gaming of the system

techtoniq committed:
Content changes per latest prototype

techtoniq committed:
Fix nav issue

techtoniq committed:
Content tweaks

techtoniq committed:
Tweak to content

techtoniq committed:
Tweak to content

techtoniq committed:
Content tweaks

techtoniq committed:
Content tweak

CodeWolf86 commented on src/SFA.DAS.ApprenticeFeedback.Web/Pages/ExitSurvey/Invalid.cshtml:
```diff
@@ -0,0 +1,7 @@
+}
+
+
+<p>apprentice feedback target id not valid for the logged in apprentice, or some other invalid scenario</p>
```
Do we still need these two files?

techtoniq committed:
Remove page that is not needed

techtoniq committed:
QF-691 prevent exit survey when apprentice not withdrawn, and prevent exit survey when already completed

techtoniq committed:
Unit tests to check QF-691

techtoniq committed:
Change unhappy path submit button caption from "Confirm and submit" to "Submit" per Slack chat with Mo and Victor

techtoniq committed:
Fix regression bug where header was being hidden from feedback survey pages too

techtoniq committed:
QF-691 fix

techtoniq committed:
Fix missing validation error message

techtoniq committed:
Fix for bug showing final completion page
